src/devtools/client/webconsole/utils/autocomplete.ts

```diff
@@ -0,0 +1,249 @@
+import * as babelParser from "@babel/parser";
+import * as t from "@babel/types";
+const { filter } = require("fuzzaldrin-plus");
+
+type ObjectPreviewProperty = {
+  value: ValueFront;
+  writable: boolean;
+  configurable: boolean;
+  enumerable: boolean;
+  name: string;
+  get?: any;
+  set?: any;
+};
+type ObjectPreview = {
+  containerEntries: any[];
+  getterValues: any[];
+  properties: ObjectPreviewProperty[];
+  prototypeId: any;
+  prototypeValue: ValueFront;
+};
+type ObjectFront = {
+  className: string;
+  objectId: string;
+  preview?: ObjectPreview;
+};
+type ValueFront = {
+  _pause: any;
+  _hasPrimitive: false;
+  _primitive?: boolean;
+  _isBigInt: boolean;
+  _isSymbol: boolean;
+  _object: ObjectFront;
+  _uninitialized: boolean;
+  _unavailable: boolean;
+  _elements: null;
+  _isMapEntry: null;
+};
+type Binding = {
+  name: string;
+  value: ValueFront;
+};
+type Scope = {
+  bindings: Binding[];
+  actor: string;
+  parent: Scope;
+  functionName?: string;
+  object?: ValueFront;
+  scopeKind: string;
+  type: "block" | string;
+};
+type FrameScope = {
+  scope: Scope;
+  pending: boolean;
+  originalScopesUnavailable: boolean;
+};
+
+function getMatchString(str: string) {
+  return str.toLowerCase().replace(/['"`]+/g, "");
+}
+function fuzzyFilter(candidates: string[], query: string): string[] {
+  if (getMatchString(query) === "") {
+    return candidates;
+  }
+  return filter(candidates, query);
+}
+
+function isIdentifier(expression: string) {
+  try {
+    const parsed = babelParser.parseExpression(expression);
+    const isIdentifier = t.isIdentifier(parsed);
+
+    return isIdentifier;
+  } catch (e) {
+    return false;
+  }
+}
+function isMemberExpression(expression: string) {
+  try {
+    const parsed = babelParser.parseExpression(expression);
+    const rv = t.isMemberExpression(parsed);
+
+    return rv;
+  } catch (e) {
+    return false;
+  }
+}
+
+function isBracketNotation(input: string) {
+  const expression = input.slice(0, input.lastIndexOf("[") + 1);
+  return isMemberExpression(expression + `"a"]`);
+}
+function isDotNotation(input: string) {
+  return isMemberExpression(input + `a`);
```

Comment:
Ah, ok since we're using the `a` thing multiple times, could we pull it out to like

```
// This stands in for a single JS token in places where we don't have a complete expression yet
const TOKEN_PLACEHOLDER = "tokenPlaceholder"
```